stylus-proc/src/macros/entrypoint.rs

```diff
@@ -0,0 +1,199 @@
+                struct_entrypoint_fn: struct_entrypoint_fn(&item.ident),
+                assert_overrides_const: assert_overrides_const(&item.ident),
+                item,
+            }),
+            _ => abort!(item, "not a struct or fn"),
+        };
+
+        Ok(Self {
+            user_entrypoint_fn: user_entrypoint_fn(kind.entrypoint_fn_name()),
+            mark_used_fn: mark_used_fn(),
+            kind,
+        })
+    }
+}
+
+impl ToTokens for Entrypoint {
+    fn to_tokens(&self, tokens: &mut TokenStream) {
+        self.kind.to_tokens(tokens);
+        self.mark_used_fn.to_tokens(tokens);
+        self.user_entrypoint_fn.to_tokens(tokens);
+    }
+}
+
+enum EntrypointKind {
+    Fn(EntrypointFn),
+    Struct(EntrypointStruct),
+}
+
+impl EntrypointKind {
+    fn entrypoint_fn_name(&self) -> Ident {
+        match self {
+            EntrypointKind::Fn(EntrypointFn { item }) => item.sig.ident.clone(),
+            EntrypointKind::Struct(EntrypointStruct { item, .. }) => {
+                let mut ident = STRUCT_ENTRYPOINT_FN.as_ident();
+                ident.set_span(item.ident.span());
+                ident
+            }
+        }
+    }
+}
+
+impl ToTokens for EntrypointKind {
+    fn to_tokens(&self, tokens: &mut TokenStream) {
+        match self {
+            EntrypointKind::Fn(inner) => inner.to_tokens(tokens),
+            EntrypointKind::Struct(inner) => inner.to_tokens(tokens),
+        }
+    }
+}
+
+struct EntrypointFn {
+    item: syn::ItemFn,
+}
+
+impl ToTokens for EntrypointFn {
+    fn to_tokens(&self, tokens: &mut TokenStream) {
+        self.item.to_tokens(tokens);
+    }
+}
+
+struct EntrypointStruct {
+    item: syn::ItemStruct,
+    top_level_storage_impl: syn::ItemImpl,
+    struct_entrypoint_fn: syn::ItemFn,
+    assert_overrides_const: syn::ItemConst,
+}
+
+impl ToTokens for EntrypointStruct {
+    fn to_tokens(&self, tokens: &mut TokenStream) {
+        self.item.to_tokens(tokens);
+        self.top_level_storage_impl.to_tokens(tokens);
+        self.struct_entrypoint_fn.to_tokens(tokens);
+        self.assert_overrides_const.to_tokens(tokens);
+    }
+}
+
+fn top_level_storage_impl(item: &syn::ItemStruct) -> syn::ItemImpl {
+    let name = &item.ident;
+    let (impl_generics, ty_generics, where_clause) = item.generics.split_for_impl();
+    parse_quote! {
+        unsafe impl #impl_generics stylus_sdk::storage::TopLevelStorage for #name #ty_generics #where_clause {}
+    }
+}
+
+fn struct_entrypoint_fn(name: &Ident) -> syn::ItemFn {
+    parse_quote! {
+        fn #STRUCT_ENTRYPOINT_FN(input: alloc::vec::Vec<u8>) -> stylus_sdk::ArbResult {
+            use stylus_sdk::{abi::Router, alloy_primitives::U256, console, hex, storage::StorageType};
+            use core::convert::TryInto;
+            use alloc::vec;
+
+            if input.len() < 4 {
+                console!("calldata too short: {}", hex::encode(input));
+                return Err(vec![]);
```

Comment:
wouldnt it be better to return a more expressive error here? Maybe for a future pr